# EpichloeSpeciesTreeProject
Full code repository for code utilised during 2019-2020 BPRC Summer Scholarship to determine a well supported species tree for _Epichloe_ fungal spp.

Scripts directory contains all scripts utilised

1. use master.sh with orthologs.py and auto_align.sh to create alignments and trees

2. check alignments with alignmentstats.py

3. import trees into R and create species and strain trees using fulltreescript.R

This repository will contain code utilised for ortholog analysis, tree production and some graphic representation production.

Note these scripts start on assumption that orthologs for each species are identified i.e. genome data is already annotated (funannotate).
